### PR TITLE
Bug/priority fault queues

### LINKED ISSUE #261: Ensure Faults are Always Handled
### Description

Currently, the queue_fault function uses queue_and_set_flag, which uses try semantics and will thus drop a fault if there is no space in the queue. We should change this where all faults are guaranteed to be handled, which either involves never making it be dropped from a queue, or changing the fault handling system to guarantee that all faults are handled. 

### Acceptance Criteria

- [ ] Queue fault doesn't allow for faults to be dropped
- [ ] It is impossible for a fault to be triggered by a thread and not make it to the fault thread
- [ ] It is guaranteed that the car faults within the time stated by the rules

### Proposed Solution

- [ ] Change fault queueing to never drop items
- [ ] Make a global fault variable that can be accessed from any rtos context, or make a global variable that can only be accessed from a task context, and make all faults only be triggered from a task context

### PR DESCRIPTION
**Pull request by fregely**
## Changes

added separate queue for low and high priority faults 

## Notes

_Any other notes go here_

## Test Cases

- Case A
- Edge case
- ...

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please reach out to your Project Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #261

### CODE REVIEW

File: Core/Src/fault.c [hunk 1]
```diff
@@ -54,70 +64,79 @@ void vFaultHandler(void *pv_params)
 	}
 
 	fault_data_t fault_data;
-	fault_handle_queue = osMessageQueueNew(FAULT_HANDLE_QUEUE_SIZE,
-					       sizeof(fault_data_t), NULL);
-
+	fault_handle_low_priority_queue = osMessageQueueNew(
+		FAULT_HANDLE_QUEUE_SIZE / 2, sizeof(fault_data_t), NULL);
+	fault_handle_low_priority_queue = osMessageQueueNew(
+		FAULT_HANDLE_QUEUE_SIZE / 2, sizeof(fault_data_t), NULL);
```
Comment (caiodasilva2005) [inline Core/Src/fault.c]: Let's keep both queues at FAULT_HANDLE_QUEUE_SIZE, no need to div 2

File: Core/Src/fault.c [hunk 2]
```diff
@@ -54,70 +64,79 @@ void vFaultHandler(void *pv_params)
...
-				break;
-			default:
-				break;
-			}
+			// Process high-priority fault
+			processFault(fault_data);
+		} else if (osMessageQueueGet(fault_handle_low_priority_queue,
+					     &fault_data, NULL,
+					     pdMS_TO_TICKS(SEND_FAULT_TIME)) ==
+			   osOK) {
```
Comment (caiodasilva2005) [inline Core/Src/fault.c]: Small change: can we maybe just do
 if (osMessageQueueGet(high_prio) == osOk ||  osMessageQueueGet(low_prio) == osOk) {
 process_fault(fault_data)
 }
 
 just think it would be a little cleaner (thats just the psuedocode for it tho)

File: Core/Inc/fault.h [hunk 3]
```diff
@@ -41,21 +41,22 @@ typedef struct {
...
- * 
+ *
  * @param pv_params NULL
  */
 void vFaultHandler(void *pv_params);
 extern osThreadId_t fault_handle;
 extern const osThreadAttr_t fault_handle_attributes;
 void clearFault(void *args);
+void processFault(fault_data_t fault_data);
 fault_sev_t getMaxSeverity();
```
Comment (caiodasilva2005) [inline Core/Inc/fault.h]: Can you change these function names from camel case to process_fault()? I know you didn't make the other functions tho lol but would be cool if you can change them

**Review by caiodasilva2005 — COMMENTED** [review]
Great work on this! Just a few changes. Also do you wanna make the defcon -> is_critical change or do we wanna do that with the next pr for the separate bitfields? Either way is cool with me

File: Core/Src/fault.c (hunk 2 above)
Comment (fregely) [reply Core/Src/fault.c]: don't think I can use OR in this case because whatever is called second in the statement will overwrite fault_data. If both a high-priority and a low-priority fault are available, the second call to osMessageQueueGet() will overwrite the first one, causing the fault to be lost.

Comment (caiodasilva2005) [reply Core/Src/fault.c]: Ah yes I see, then you can disregard that comment

**Review by caiodasilva2005 — CHANGES_REQUESTED** [review]
Looks great! Last thing, can you just revert changes to the clang format file and then we can merge

**Comment by jr1221** [comment]
This should get closed right?

**Comment by caiodasilva2005** [comment]
Yeah gonna close this since other pr that branched off of this one is looking good and pretty much ready to merge

**Comment by caiodasilva2005** [comment]
Yeah gonna close this since other pr that branched off of this one is looking good and almost ready to merge
